[FLINK-11908][table] Port window classes into flink-api-java

hequn8128 opened the pull request:
## What is the purpose of the change

This pull request ports window classes into flink-api-java module.


## Brief change log

  - Convert Tumble/Slide/Session/Over related classes to interface and rename the original class from XXX to XXXImpl. For example, add a TumbleWithSize interface and rename the original TumbleWithSize to TumbleWithSizeImpl. We can't port TumbleWithSize directly into the api-java module because ExpressionParser is used in it.
  - Deprecate `def window(window: Window): WindowedTable` and add `def window(window: GroupWindow): GroupWindowedTable`
  - Deprecate class WindowedTable and add GroupWindowedTable
  - Add test case, such as WindowValidationTest, to test failures for the creation of window. 


## Verifying this change

This change is already covered by existing Window IT/UT test cases.
Furthermore, WindowValidationTest has been add to test failures for the creation of window. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

sunjincheng121 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/GroupWindow.java:
```diff
@@ -0,0 +1,41 @@
+ * <p>For finite batch tables, group windows provide shortcuts for time-based groupBy.
+ *
+ * <p>Note: {@link Window} is temporally used as the father class of {@link GroupWindow} for the
+ * sake of compatibility. It will be removed later.
```
Here just a suggestion, how about we open a new PR for Deprecated the Window in release-1.8, since we should create a new RC2 for release 1.8. If we do not do that the Window will keep existing for almost half a year. And I'll create the Jira FLINK-11918, and link to release-1.8 vote mail thread, ask RM's options. If all of you do not agree, I'll close the JIRA, otherwise, we can open the new PR for Deprecated the window.

sunjincheng121 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Tumble.java:
```diff
@@ -0,0 +1,88 @@
+ * elements in 5 minutes intervals.
+ *
+ * <p>Java users should use the methods with parameters of String type, while Scala users should use
+ * the methods with parameters of Expression type.
```
Same as above.

sunjincheng121 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -1092,10 +1094,35 @@ class Table(
+  @Deprecated
+  @deprecated(
+    "This method will be removed. Use table.window(window: GroupWindow) instead",
+    "1.9.0")
```
It depends on FLINK-11918

sunjincheng121 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala:
```diff
@@ -333,7 +348,8 @@ class SlideWithSize(size: Expression) {
     */
-  def every(slide: Expression): SlideWithSizeAndSlide = new SlideWithSizeAndSlide(size, slide)
+  def every(slide: Expression): SlideWithSizeAndSlide =
+    new SlideWithSizeAndSlideImpl(size, slide)
```
Unnecessary change？

sunjincheng121 commented:
@flinkbot approve description
@flinkbot approve consensus

hequn8128 commented:
@sunjincheng121 Thanks a lot for your review. I have addressed all your comments.

@twalthr @sunjincheng121 I have also rebased to the deprecating window pr and delete the deprecated methods and classes in this one. Would be great if you can take a look. 

Best, Hequn

hequn8128 commented:
Update, rebase to the latest deprecating window [PR](https://github.com/apache/flink/pull/7985) to unblock review. The updates contain two commits. One is the PR of deprecating window, the other one is the PR of the current issue.

hequn8128 commented:
Rebase to the latest flink/master

hequn8128 commented:
@twalthr Sorry. I missed the important messages from you. Convert the ExpressionParser into an interface is a good solution. So we can port TumbleWithSize, TumbleWithSizeOnTime, etc. directly into api-java. This also performs a reflection step at just one location. 

Good idea! 
I will try to update the PR asap.

Best, Hequn

hequn8128 commented:
@twalthr @sunjincheng121 I have updated the PR. Would be great if you can take another look. 

Some changes are explained by the followings:
- To make `ExpressionParser.parseExpression()` available, the `ExpressionParser` contains two static methods: `parseExpression()` and `parseExpressionList()`.
- As static methods cannot be overridden, I added a `PlannerExpressionParser` interface. The `PlannerExpressionParser` contains two non-static expression parser methods and are overridden by `PlannerExpressionParserImpl`. So once we get a `PlannerExpressionParserImpl` instance through refection, we can call parseExpression() in `PlannerExpressionParserImpl`.
- Furthermore, I add a companion class `PlannerExpressionParserImpl`. This is because we can't new object directly.
- Both `ExpressionParser` and `PlannerExpressionParser` in api-java are marked as Internal class.

Best, 
Hequn

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowPartitionedOrdered.java:
```diff
@@ -0,0 +1,86 @@
+			alias,
+			partitionBy,
+			orderBy,
+			ExpressionParser.parseExpression("UNBOUNDED_RANGE"),
```
Don't introduce string constants at an arbitrary code location. Instead, either move it to the top of a class as a static final field or in this case use an already existing constant in `BuiltInFunctionDefinitions`.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowPartitioned.java:
```diff
@@ -0,0 +1,73 @@
+	/** Defines a partitioning of the input on one or more attributes. */
+	private final List<Expression> partitionBy;
+
+	public OverWindowPartitioned(List<Expression> partitionBy) {
```
We can use a default visibility for intermediate API classes. Such that users can not accidentally use the wrong methods.

twalthr commented:
@flinkbot approve all

hequn8128 commented:
@twalthr Thanks for your valuable suggestions. I will move on the next PR(convert Table to the interface).
